modules/imgproc/src/color_lab.cpp

```diff
@@ -424,134 +289,133 @@ template<typename _Tp> struct RGB2XYZ_i
+            v_zip(sb1, sg1, bg[2], bg[3]);
+            v_zip(sr0, vdescale, rd[0], rd[1]);
+            v_zip(sr1, vdescale, rd[2], rd[3]);
+
+            v_uint32 vx[4], vy[4], vz[4];
+            for(int j = 0; j < 4; j++)
             {
-                uint8x8x4_t v_src = vld4_u8(src);
-                v_src16.val[0] = vmovl_u8(v_src.val[0]);
-                v_src16.val[1] = vmovl_u8(v_src.val[1]);
-                v_src16.val[2] = vmovl_u8(v_src.val[2]);
+                vx[j] = v_reinterpret_as_u32(v_dotprod(bg[j], cxbg) + v_dotprod(rd[j], cxr1)) >> shift;
+                vy[j] = v_reinterpret_as_u32(v_dotprod(bg[j], cybg) + v_dotprod(rd[j], cyr1)) >> shift;
+                vz[j] = v_reinterpret_as_u32(v_dotprod(bg[j], czbg) + v_dotprod(rd[j], czr1)) >> shift;
             }
 
-            uint16x4_t v_s0 = vget_low_u16(v_src16.val[0]),
-                       v_s1 = vget_low_u16(v_src16.val[1]),
-                       v_s2 = vget_low_u16(v_src16.val[2]);
-
-            uint32x4_t v_X0 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c0), v_s1, v_c1), v_s2, v_c2);
-            uint32x4_t v_Y0 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c3), v_s1, v_c4), v_s2, v_c5);
-            uint32x4_t v_Z0 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c6), v_s1, v_c7), v_s2, v_c8);
-            v_X0 = vshrq_n_u32(vaddq_u32(v_X0, v_delta), xyz_shift);
-            v_Y0 = vshrq_n_u32(vaddq_u32(v_Y0, v_delta), xyz_shift);
-            v_Z0 = vshrq_n_u32(vaddq_u32(v_Z0, v_delta), xyz_shift);
-
-            v_s0 = vget_high_u16(v_src16.val[0]),
-            v_s1 = vget_high_u16(v_src16.val[1]),
-            v_s2 = vget_high_u16(v_src16.val[2]);
-
-            uint32x4_t v_X1 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c0), v_s1, v_c1), v_s2, v_c2);
-            uint32x4_t v_Y1 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c3), v_s1, v_c4), v_s2, v_c5);
-            uint32x4_t v_Z1 = vmlal_u16(vmlal_u16(vmull_u16(v_s0, v_c6), v_s1, v_c7), v_s2, v_c8);
-            v_X1 = vshrq_n_u32(vaddq_u32(v_X1, v_delta), xyz_shift);
-            v_Y1 = vshrq_n_u32(vaddq_u32(v_Y1, v_delta), xyz_shift);
-            v_Z1 = vshrq_n_u32(vaddq_u32(v_Z1, v_delta), xyz_shift);
-
-            v_dst.val[0] = vqmovn_u16(vcombine_u16(vmovn_u32(v_X0), vmovn_u32(v_X1)));
-            v_dst.val[1] = vqmovn_u16(vcombine_u16(vmovn_u32(v_Y0), vmovn_u32(v_Y1)));
-            v_dst.val[2] = vqmovn_u16(vcombine_u16(vmovn_u32(v_Z0), vmovn_u32(v_Z1)));
-
-            vst3_u8(dst + i, v_dst);
+            v_uint16 x0, x1, y0, y1, z0, z1;
+            x0 = v_pack(vx[0], vx[1]);
+            x1 = v_pack(vx[2], vx[3]);
+            y0 = v_pack(vy[0], vy[1]);
+            y1 = v_pack(vy[2], vy[3]);
+            z0 = v_pack(vz[0], vz[1]);
+            z1 = v_pack(vz[2], vz[3]);
+
+            v_uint8 x, y, z;
+            x = v_pack(x0, x1);
+            y = v_pack(y0, y1);
+            z = v_pack(z0, z1);
+
+            v_store_interleave(dst, x, y, z);
         }
+        vx_cleanup();
```

Comment:
`vx_cleanup()` in the middle of function makes no sense (useless).
Compiler is still free to use AVX instructions for the loop below (if AVX code generation is enabled, then it is enabled for the whole file).

Reply:
Does this mean that `vx_cleanup()` should be used only in a function's end?

Reply:
It is needed of SSE/AVX mode switches.
This may happen only in the end or the beginning of the AVX-optimized function.

`CV_INSTRUMENT_REGION();` handles both cases (via `CV_AVX_GUARD`).